File: BirthdayBot/UserInterface/UserCommands.vb
Imports System.Text.RegularExpressions
Imports Discord.WebSocket

Class UserCommands
    Inherits CommandsCommon

    Public Overrides ReadOnly Property Commands As IEnumerable(Of (String, CommandHandler))
        Get
            Return New List(Of (String, CommandHandler)) From {
                ("set", AddressOf CmdSet),
                ("zone", AddressOf CmdZone),
                ("remove", AddressOf CmdRemove),
                ("when", AddressOf CmdWhen)
            }
        End Get
    End Property

    Sub New(inst As BirthdayBot, db As Configuration)
        MyBase.New(inst, db)
    End Sub

    ''' <summary>
    ''' Parses date parameter. Strictly takes dd-MMM or MMM-dd only. Eliminates ambiguity over dd/mm vs mm/dd.
    ''' </summary>
    ''' <returns>Tuple: month, day</returns>
    ''' <exception cref="FormatException">Thrown for any parsing issue. Reason is expected to be sent to Discord as-is.</exception>
    Private Function ParseDate(dateInput As String) As (Integer, Integer)
        ' Not using DateTime.Parse. Setting it up is rather complicated, and it's probably case sensitive.
        ' Admittedly, doing it the way it's being done here probably isn't any better.
        Dim m = Regex.Match(dateInput, "^(?<day>\d{1,2})-(?<month>[A-Za-z]{3})$")
        If Not m.Success Then
            ' Flip the fields around, try again
            m = Regex.Match(dateInput, "^(?<month>[A-Za-z]{3})-(?<day>\d{1,2})$")
            If Not m.Success Then Throw New FormatException(GenericError)
        End If
        Dim day As Integer
        Try
            day = Integer.Parse(m.Groups("day").Value)
        Catch ex As FormatException
            Throw New FormatException(GenericError)
        End Try
        Dim monthVal = m.Groups("month").Value
        Dim month As Integer
        Dim dayUpper = 31 ' upper day of month check
        Select Case monthVal.ToLower()
            Case "jan"
                month = 1
            Case "feb"
                month = 2
                dayUpper = 29
            Case "mar"
                month = 3
            Case "apr"
                month = 4
                dayUpper = 30
            Case "may"
                month = 5
            Case "jun"
                month = 6
                dayUpper = 30
            Case "jul"
                month = 7
            Case "aug"
                month = 8
            Case "sep"
                month = 9
                dayUpper = 30
            Case "oct"
                month = 10
            Case "nov"
                month = 11
                dayUpper = 30
            Case "dec"
                month = 12
            Case Else
                Throw New FormatException(":x: Invalid month name. Use a three-letter month abbreviation.")
        End Select
        If day = 0 Or day > dayUpper Then Throw New FormatException(":x: The date you specified is not a valid calendar date.")

        Return (month, day)
    End Function

    Private Async Function CmdSet(param As String(), reqChannel As SocketTextChannel, reqUser As SocketGuildUser) As Task
        ' Requires one parameter. Optionally two.
        If param.Count < 2 Or param.Count > 3 Then
            Await reqChannel.SendMessageAsync(GenericError)
            Return
        End If

        Dim bmonth, bday As Integer
        Dim btz As String = Nothing
        Try
            Dim res = ParseDate(param(1))
            bmonth = res.Item1
            bday = res.Item2
            If param.Length = 3 Then
                btz = ParseTimeZone(param(2))
            End If
        Catch ex As FormatException
            ' Our parse methods' FormatException has its message to send out to Discord.
            reqChannel.SendMessageAsync(ex.Message).Wait()
            Return
        End Try

        ' Parsing successful. Update user information.
        Dim known As Boolean ' Extra detail: Bot's response changes if the user was previously unknown.
        Try
            Dim user = Instance.GuildCache(reqChannel.Guild.Id).GetUser(reqUser.Id)
            known = user.IsKnown
            Await user.UpdateAsync(bmonth, bday, btz, BotConfig.DatabaseSettings)
        Catch ex As Exception
            Log("Error", ex.ToString())
            reqChannel.SendMessageAsync(":x: An unknown error occurred. The bot owner has been notified.").Wait()
            Return
        End Try
        If known Then
            Await reqChannel.SendMessageAsync(":white_check_mark: Your information has been updated.")
        Else
            Await reqChannel.SendMessageAsync(":white_check_mark: Your birthday has been recorded.")
        End If
    End Function

    Private Async Function CmdZone(param As String(), reqChannel As SocketTextChannel, reqUser As SocketGuildUser) As Task
        If param.Count <> 2 Then
            Await reqChannel.SendMessageAsync(GenericError)
            Return
        End If

        Dim btz As String = Nothing
        Dim user = Instance.GuildCache(reqChannel.Guild.Id).GetUser(reqUser.Id)
        If Not user.IsKnown Then
            Await reqChannel.SendMessageAsync(":x: Can't set your time zone if your birth date isn't registered.")
            Return
        End If

        Try
            btz = ParseTimeZone(param(1))
        Catch ex As Exception
            reqChannel.SendMessageAsync(ex.Message).Wait()
            Return
        End Try
        Await user.UpdateAsync(user.BirthMonth, user.BirthDay, btz, BotConfig.DatabaseSettings)

        Await reqChannel.SendMessageAsync($":white_check_mark: Your time zone has been updated to **{btz}**.")
    End Function

    Private Async Function CmdRemove(param As String(), reqChannel As SocketTextChannel, reqUser As SocketGuildUser) As Task
        ' Parameter count check
        If param.Count <> 1 Then
            Await reqChannel.SendMessageAsync(ExpectedNoParametersError)
            Return
        End If

        ' Extra detail: Send a notification if the user isn't actually known by the bot.
        Dim known As Boolean
        Dim g = Instance.GuildCache(reqChannel.Guild.Id)
        known = g.GetUser(reqUser.Id).IsKnown
        ' Delete database and cache entry
        Await g.DeleteUserAsync(reqUser.Id)
        If Not known Then
            Await reqChannel.SendMessageAsync(":white_check_mark: I don't have your information. Nothing to remove.")
        Else
            Await reqChannel.SendMessageAsync(":white_check_mark: Your information has been removed.")
        End If
    End Function

    Private Async Function CmdWhen(param As String(), reqChannel As SocketTextChannel, reqUser As SocketGuildUser) As Task
        ' Requires a parameter
        If param.Count = 1 Then
            Await reqChannel.SendMessageAsync(GenericError)
            Return
        End If

        Dim search = param(1)
        If param.Count = 3 Then
            ' param maxes out at 3 values. param(2) might contain part of the search string (if name has a space)
            search += " " + param(2)
        End If

        Dim searchTarget As SocketGuildUser = Nothing

        Dim searchId As ULong = 0
        If Not TryGetUserId(search, searchId) Then ' ID lookup
            ' name lookup without discriminator
            For Each searchuser In reqChannel.Guild.Users
                If String.Equals(search, searchuser.Username, StringComparison.OrdinalIgnoreCase) Then
                    searchTarget = searchuser
                    Exit For
                End If
            Next
        Else
            searchTarget = reqChannel.Guild.GetUser(searchId)
        End If
        If searchTarget Is Nothing Then
            Await reqChannel.SendMessageAsync(BadUserError)
            Return
        End If

        Dim users = Instance.GuildCache(reqChannel.Guild.Id).Users
        Dim searchTargetData = users.FirstOrDefault(Function(u) u.UserId = searchTarget.Id)
        If searchTargetData Is Nothing Then
            Await reqChannel.SendMessageAsync("The given user does not exist or has not set a birthday.")
            Return
        End If

        With searchTargetData
            Await reqChannel.SendMessageAsync(FormatName(searchTarget, False) + ": " +
                                          $"`{ .BirthDay.ToString("00")}-{Common.MonthNames(.BirthMonth)}`" +
                                          If(.TimeZone Is Nothing, "", $" - `{ .TimeZone}`"))
        End With
    End Function
End Class

File: TextCommands/UserCommands.cs
#pragma warning disable CS0618
using BirthdayBot.Data;
using System.Text.RegularExpressions;

namespace BirthdayBot.TextCommands;

internal class UserCommands : CommandsCommon {
    public UserCommands(Configuration db) : base(db) { }

    public override IEnumerable<(string, CommandHandler)> Commands
        => new List<(string, CommandHandler)>()
        {
                ("set", CmdSet),
                ("zone", CmdZone),
                ("remove", CmdRemove)
        };

    #region Date parsing
    const string FormatError = ":x: Unrecognized date format. The following formats are accepted, as examples: "
            + "`15-jan`, `jan-15`, `15 jan`, `jan 15`, `15 January`, `January 15`.";

    private static readonly Regex DateParse1 = new(@"^(?<day>\d{1,2})[ -](?<month>[A-Za-z]+)$", RegexOptions.Compiled);
    private static readonly Regex DateParse2 = new(@"^(?<month>[A-Za-z]+)[ -](?<day>\d{1,2})$", RegexOptions.Compiled);

    /// <summary>
    /// Parses a date input.
    /// </summary>
    /// <returns>Tuple: month, day</returns>
    /// <exception cref="FormatException">
    /// Thrown for any parsing issue. Reason is expected to be sent to Discord as-is.
    /// </exception>
    private static (int, int) ParseDate(string dateInput) {
        var m = DateParse1.Match(dateInput);
        if (!m.Success) {
            // Flip the fields around, try again
            m = DateParse2.Match(dateInput);
            if (!m.Success) throw new FormatException(FormatError);
        }

        int day, month;
        string monthVal;
        try {
            day = int.Parse(m.Groups["day"].Value);
        } catch (FormatException) {
            throw new Exception(FormatError);
        }
        monthVal = m.Groups["month"].Value;

        int dayUpper; // upper day of month check
        (month, dayUpper) = GetMonth(monthVal);

        if (day == 0 || day > dayUpper) throw new FormatException(":x: The date you specified is not a valid calendar date.");

        return (month, day);
    }

    /// <summary>
    /// Returns information for a given month input.
    /// </summary>
    /// <param name="input"></param>
    /// <returns>Tuple: Month value, upper limit of days in the month</returns>
    /// <exception cref="FormatException">
    /// Thrown on error. Send out to Discord as-is.
    /// </exception>
    private static (int, int) GetMonth(string input) {
        return input.ToLower() switch {
            "jan" or "january" => (1, 31),
            "feb" or "february" => (2, 29),
            "mar" or "march" => (3, 31),
            "apr" or "april" => (4, 30),
            "may" => (5, 31),
            "jun" or "june" => (6, 30),
            "jul" or "july" => (7, 31),
            "aug" or "august" => (8, 31),
            "sep" or "september" => (9, 30),
            "oct" or "october" => (10, 31),
            "nov" or "november" => (11, 30),
            "dec" or "december" => (12, 31),
            _ => throw new FormatException($":x: Can't determine month name `{input}`. Check your spelling and try again."),
        };
    }
    #endregion

    #region Documentation
    public static readonly CommandDocumentation DocSet =
        new(new string[] { "set (date)" }, "Registers your birth month and day.",
            $"`{CommandPrefix}set jan-31`, `{CommandPrefix}set 15 may`.");
    public static readonly CommandDocumentation DocZone =
        new(new string[] { "zone (zone)" }, "Sets your local time zone. "
            + $"See also `{CommandPrefix}help-tzdata`.", null);
    public static readonly CommandDocumentation DocRemove =
        new(new string[] { "remove" }, "Removes your birthday information from this bot.", null);
    #endregion

    private async Task CmdSet(ShardInstance instance, GuildConfiguration gconf,
                              string[] param, SocketTextChannel reqChannel, SocketGuildUser reqUser) {
        if (param.Length < 2) {
            await reqChannel.SendMessageAsync(ParameterError, embed: DocSet.UsageEmbed).ConfigureAwait(false);
            return;
        }

        // Date format accepts spaces. Must coalesce parameters to a single string.
        var fullinput = "";
        foreach (var p in param[1..]) fullinput += " " + p;
        fullinput = fullinput[1..]; // trim leading space

        int bmonth, bday;
        try {
            (bmonth, bday) = ParseDate(fullinput);
        } catch (FormatException ex) {
            // Our parse method's FormatException has its message to send out to Discord.
            reqChannel.SendMessageAsync(ex.Message, embed: DocSet.UsageEmbed).Wait();
            return;
        }

        // Parsing successful. Update user information.
        bool known; // Extra detail: Bot's response changes if the user was previously unknown.
        try {
            var user = await GuildUserConfiguration.LoadAsync(gconf.GuildId, reqUser.Id).ConfigureAwait(false);
            known = user.IsKnown;
            await user.UpdateAsync(bmonth, bday, user.TimeZone).ConfigureAwait(false);
        } catch (Exception ex) {
            Program.Log("Error", ex.ToString());
            reqChannel.SendMessageAsync(ShardInstance.InternalError).Wait();
            return;
        }
        await reqChannel.SendMessageAsync($":white_check_mark: Your birthday has been { (known ? "updated" : "recorded") }.")
                .ConfigureAwait(false);
    }

    private async Task CmdZone(ShardInstance instance, GuildConfiguration gconf,
                               string[] param, SocketTextChannel reqChannel, SocketGuildUser reqUser) {
        if (param.Length != 2) {
            await reqChannel.SendMessageAsync(ParameterError, embed: DocZone.UsageEmbed).ConfigureAwait(false);
            return;
        }

        var user = await GuildUserConfiguration.LoadAsync(gconf.GuildId, reqUser.Id).ConfigureAwait(false);
        if (!user.IsKnown) {
            await reqChannel.SendMessageAsync(":x: You may only update your time zone when you have a birthday registered."
                + $" Refer to the `{CommandPrefix}set` command.", embed: DocZone.UsageEmbed)
                .ConfigureAwait(false);
            return;
        }

        string btz;
        try {
            btz = ParseTimeZone(param[1]);
        } catch (Exception ex) {
            reqChannel.SendMessageAsync(ex.Message, embed: DocZone.UsageEmbed).Wait();
            return;
        }
        await user.UpdateAsync(user.BirthMonth, user.BirthDay, btz).ConfigureAwait(false);

        await reqChannel.SendMessageAsync($":white_check_mark: Your time zone has been updated to **{btz}**.")
            .ConfigureAwait(false);
    }

    private async Task CmdRemove(ShardInstance instance, GuildConfiguration gconf,
                                 string[] param, SocketTextChannel reqChannel, SocketGuildUser reqUser) {
        // Parameter count check
        if (param.Length != 1) {
            await reqChannel.SendMessageAsync(NoParameterError, embed: DocRemove.UsageEmbed).ConfigureAwait(false);
            return;
        }

        // Extra detail: Send a notification if the user isn't actually known by the bot.
        bool known;
        var u = await GuildUserConfiguration.LoadAsync(gconf.GuildId, reqUser.Id).ConfigureAwait(false);
        known = u.IsKnown;
        await u.DeleteAsync().ConfigureAwait(false);
        if (!known) {
            await reqChannel.SendMessageAsync(":white_check_mark: This bot already does not contain your information.")
                .ConfigureAwait(false);
        } else {
            await reqChannel.SendMessageAsync(":white_check_mark: Your information has been removed.")
                .ConfigureAwait(false);
        }
    }
}
